### DISCUSSION TITLE
Show updated data in Sibling Component in React using subscribe.

### DISCUSSION THREAD

**Discussion by deadcoder0904** [pmndrs/valtio#300] [Q&A]
I'm trying to use `subscribe` to show the updated data in `valtio` but confused where to use it & how to use it.

I tried searching on [Sourcegraph](https://sourcegraph.com/search?q=import+%7B+subscribe+%7D+from+%27valtio%27&patternType=literal) to see code samples but couldn't find any about using subscribe in React.

My application looks like this:

<img width="1440" alt="Screenshot 2021-12-22 at 4 11 04 PM" src="https://user-images.githubusercontent.com/16436270/147080133-4f380d7f-b084-4814-a37a-d4f1245fd0ca.png">

On the right side, I type stuff & on the left side, it should update using subscribe.

This is relatively simple but I'm unable to find a way to do it.

I did put the `subscribe` under `getDocsById` which gets called when I click on the list on the left.

```ts
const getDocsById = (id: string) => {
    if (docs) {
      const el = docs.find((item) => item?.id === id)
      subscribe(state.doc.text, () =>
        console.log('state.doc.text has changed to', state.doc.text)
      )

      if (el) {
        state.setDoc({
          id,
          text: el.text,
        })
      }
    }
  }
```

However, it doesn't log anything to the console. I'm confused about where should I put subscribe & how should I update it in the list?

Full reproduction on this commit → https://github.com/deadcoder0904/docs-autosync/tree/cba10738b549d85aba9392f9a8e6d37db6af4725

**Comment by dai-shi** [comment]
Does this wiki page help?
https://github.com/pmndrs/valtio/wiki/How-to-avoid-rerenders-manually#subscribe-and-set-local-state-conditionally

**Reply by deadcoder0904** [reply]
yes, that does make sense but i'm still unable to figure out a way to update the selected item as i'm looping over all of them so i'm not sure how to change it?

furthermore, should i use `state` or `snap` for subscribing? in one thread, you mentioned using `snap` for reading & `state` for writing. what should i use for `subscribe`? for eg, i had to use `state` for `textarea`'s value to track change because `snap` wouldn't work.

**Reply by dai-shi** [reply]
> yes, that does make sense but i'm still unable to figure out a way to update the selected item as i'm looping over all of them so i'm not sure how to change it?

sorry, I didn't have time to look into your repro yet. you should subscribe just once and unsubscribe if no longer needed.

> furthermore, should i use `state` or `snap` for subscribing?

`state`.

> in one thread, you mentioned using `snap` for reading & `state` for writing.

this wasn't precise wording. `snap` should be used in render. `state` for everything else.
we are still looking for better wording for this. see also: https://github.com/pmndrs/eslint-plugin-valtio

> for eg, i had to use `state` for `textarea`'s value to track change because `snap` wouldn't work.

that's weird. `state` isn't reactive, so it couldn't work. `snap` should be for textarea's value, but in this case you probably need `{ sync: true }`. This isn't very nice, but only the solution for now.

**Reply by deadcoder0904** [reply]
> sorry, I didn't have time to look into your repro yet. you should subscribe just once and unsubscribe if no longer needed.

can you take a look? i understood that from the `counter` example how to do it but i dont get how i should subscribe to the right item on the list when the `textarea` changes.

im confused bcz idk if i should keep a copy of `docs` which is the list in the store too to find the current item selected or not? not sure if this is related to `valtio` though. 

i couldn't get `eslint-plugin-valtio` to work so i dropped it :(

oops sorry, i'm using `snap` & it's working just fine. `state` works too but i guess it calls a function i have on every keystroke.

**Reply by dai-shi** [reply]
> can you take a look?

Hmmm, I'm not sure where to look.

This looks fine:
https://github.com/deadcoder0904/docs-autosync/blob/cba10738b549d85aba9392f9a8e6d37db6af4725/store/index.ts

> i understood that from the `counter` example how to do it but i dont get how i should subscribe to the right item on the list when the `textarea` changes.

I'm not sure what you mean by subscribe when the `textarea` changes.
In react, subscribe should usually be done in useEffect.

> im confused bcz idk if i should keep a copy of `docs` which is the list in the store too to find the current item selected or not? not sure if this is related to `valtio` though.

I guess you would want to provide a **minimal** example to explain your use case.

> i couldn't get `eslint-plugin-valtio` to work so i dropped it :(

that's unfortunate. If you can, please open an issue there.

> oops sorry, i'm using `snap` & it's working just fine. `state` works too but i guess it calls a function i have on every keystroke.

yeah, i understand the confusion.

**Reply by deadcoder0904** [reply]
i didn't even need subscribe i guess.

i just used `snap.doc.text` on the other component for it to be reactive. i'm wondering what's the point of subscribe then if i can use `snap.doc.text`?

**Reply by dai-shi** [reply]
Right, for react users, `proxy` and `useSnapshot` cover most use cases.
`subscribe` and `snapshot` are internal functions for rare use cases.

**Reply by deadcoder0904** [reply]
ah, thanks @dai-shi, amazing work on `valtio`. it simply is the best. can you tell me since you probably use it in production: do you use `state.setDoc` or mutate directly on `state.doc.id` & `state.doc.text`?

and is `yjs-valtio` similar on surface to [syncedstore](https://syncedstore.org/) in terms of features? i probably want to use it in production & `yjs-valtio` probably released at the right time :)

**Reply by dai-shi** [reply]
I prefer this: https://github.com/pmndrs/valtio/wiki/How-to-organize-actions#action-functions-defined-in-module

Originally, I wanted to have a valtio/yjs binding in syncedstore (a different name at that time). But as we discussed, I found the concept is a little different, and I thought the API I wanted wasn't possible. So, `valtio-yjs` was born. See: https://github.com/YousefED/SyncedStore/issues/5
